Implement createNodeMock for ReactTestRenderer

Aweary opened the pull request:
This lets users pass in an optional `testOptions` object as the second argument to `ReactTestRenderer.create` that allows for custom mocking behavior. As of now that only includes `createNodeMock`, which accepts the current element and should return a mock ref object.

The `testOptions` is stored on the component instance so it doesn't leak between instances.

cc @gaearon @spicyj

Aweary commented on src/renderers/testing/ReactTestMount.js:
```diff
@@ -20,6 +20,10 @@ var getHostComponentFromComposite = require('getHostComponentFromComposite');
 var invariant = require('invariant');
 
+type TestRendererMockConfig = {
+  getMockRef: (element: ReactElement) => Object,
```
Not sure how this would be typed, is there a better way to type an object that looks like a DOM node?

gaearon replied:
Looks fine to me.

gaearon commented on src/renderers/testing/ReactTestRenderer.js:
```diff
@@ -62,9 +63,13 @@ ReactTestComponent.prototype.receiveComponent = function(
+  if (this._mockConfig) {
+    var { getMockRef } = this._mockConfig;
+    if (getMockRef) {
+      return getMockRef(this._currentElement);
```
Can we make a default `mockConfig` implementation and always assume `mockConfig` exists?

Aweary replied:
Should we have a default `getMockRef` implementation  as well and just return `this._mockConfig.getMockRef(this._currentElement)`?  Or just keep that check

gaearon replied:
Yea, default impl is what I meant.

gaearon commented on src/renderers/testing/__tests__/ReactTestRenderer-test.js:
```diff
@@ -203,7 +203,24 @@ describe('ReactTestRenderer', function() {
+    expect(log).toEqual([{}]);
+  });
+
+  it('allows an optional getMockRef function', function() {
```
Let’s also test that we receive the correct element.

Aweary commented:
FYI this has some issues with refs rendered by nested components, still working that out

Aweary commented:
Hmm, I'm not sure how I can efficiently provide access to the `mockConfig` to children at an arbitrary depth. I can walk the render tree and attach it to each instance, but that's not very efficient.

gaearon commented:
I haven't looked at the diff in detail but I think you can make it have a custom transaction (like server rendering does). Then you can read the config from the transaction since it's threaded all the way down.

An alternative is to put context on TopLevelWrapper but this is probably more hacky.

Aweary commented:
Yeah I tried TopLevelWrapper first, but it's not actually set on all instances. It only seems to be set on instances at the top of the render tree. I'll give the transaction approach a try.

gaearon commented:
Context should propagate downwards too so it must be something else. Regardless transaction seems to be a better approach.

Aweary commented:
For `_topLevelWrapper` you can [see in ReactTestMount](https://github.com/facebook/react/blob/master/src/renderers/testing/ReactTestMount.js#L56) that it's only set on the top-level `_renderedComponent`. So that might just be another separate issue we should address.

gaearon commented:
Does `_topLevelWrapper` field matter? I don't think that field is important for context propagation. Context is passed down through ReactCompositeComponent and afaik this should work any number of composites down.

Aweary commented:
Oh, no, sorry I was just responding to

> An alternative is to put context on TopLevelWrapper but this is probably more hacky.

Aweary commented:
One issue I see with trying to use the transaction to propagate the mock config, is that `ReactTestComponent.prototype.getPublicInstance` doesn't seem to have access to any transaction object.

gaearon commented:
Let's pass it in there?

gaearon commented:
Actually it should be easier to save mock config from transaction in mountComponent as an instance field. Then access it anytime. Would that work?

Aweary commented:
> Let's pass it in there?

`ReactRef` doesn't currently have access to `transaction`, but I suppose it could be threaded in from `ReactReconcilier`.

> Actually it should be easier to save mock config from transaction in mountComponent as an instance field. Then access it anytime. Would that work?

Does it make sense to use the `transaction` as a host for instance-specific config? To make this work we'd have to call `injectReconcileTransaction` every time `ReactTestRenderer.create` is called, since `mockConfig` implementation can diff between calls. 

Also, Is there a guarantee that `transaction` passed to `mountComponent` will always be the transaction type that we inject in `create`? If we have sequential `create` calls it seems like we may have a potential race condition where the next transaction type is injected before the first's `mountComponent` method is called.

Global injection makes this tricky 😄

gaearon commented:
Maybe I misunderstand how it works but I thought transaction is created on every root mount. For example server rendering creates a transaction (passing `makeStaticMarkup`, a config just like `getMockRef`) on every `renderToStringImpl` call. I don't see any injection being done there, transaction is just passed as an argument to root reconciler call. Can we do the same here?

Aweary commented:
I'm probably the one misunderstanding, but from what I can see it injects the transaction class in `ReactTestRenderer`:

```js
ReactUpdates.injection.injectReconcileTransaction(
  ReactTestReconcileTransaction
);
```

Then the actual `transaction` object is constructed from the object pool every time it's needed, which is on mount [and on update](https://github.com/facebook/react/blob/master/src/renderers/testing/ReactTestMount.js#L97-L106)

```js
  var transaction = ReactUpdates.ReactReconcileTransaction.getPooled();
  var image = transaction.perform(
    mountComponentIntoNode,
    null,
    componentInstance,
    transaction
  );
  ReactUpdates.ReactReconcileTransaction.release(transaction);
```

**but** the way `ReactServerRenderingTransaction` works with `makeStaticMarkup` is that `makeStaticMarkup` is actually passed to the `getPooled` call every time the transaction is created. So even though the transaction is recreated from an injected global class, it gets that call-specific config every time. So I think we can just do the same here! I'll give that a try.

gaearon commented:
Dynamic injection is a mess. Sorry.

Aweary commented:
Yes it is, and no sorry needed 🤘 It's at the very least a fun challenge

cpojer commented:
I was wondering if we could have a top level getMockRef per test renderer instance that would get arguments based on the ref, like the current element and the name of the expected ref? For example it could be called with the instance of an `<Input />` component that renders down to a `<input ref="text" />` component. getMockRef would then be called with `getMockRef(InputInstance, 'text)' and would be called once per ref in the tree. I'm not sure if this makes sense, just figured I'll throw the idea out there.

Aweary commented:
> I was wondering if we could have a top level getMockRef per test renderer instance that would get arguments based on the ref, like the current element and the name of the expected ref? For example it could be called with the instance of an <Input /> component that renders down to a <input ref="text" /> component. getMockRef would then be called with `getMockRef(InputInstance, 'text)' and would be called once per ref in the tree. I'm not sure if this makes sense, just figured I'll throw the idea out there.

I think that's what we're working on here, if I'm following. `getMockRef` is called when `getPublicInstance` is supposed to be called (which is what's passed to the `ref` function or attached as the `ref` for legacy refs). We'll pass in the current element to `getMockRef` so users can decide what kind of mock instance to return. Does that sound in line with what you're thinking?

cpojer commented:
Yeah exactly! But how can I decide which mock ref to use if there are two different refs used in the same component?

Aweary commented:
Passing in the current element should let you return different mocks for different element types. Is there a use case for returning different mocks for the same element type (e.g., two refs on two divs)? If not then the current element should be sufficient

cpojer commented:
Oh I see, this is an element, not a component.

Aweary commented:
Even if it was a component, the 'type' attribute should tell you which component it is 👍

Aweary commented:
So, looking into attaching the `mockConfig` to the transaction: the one issue I'm running into is that accessing the `transaction` in `ReactRef.attachRefs` isn't as straightforward as I'd hoped. 

`ReactRef.attachRefs` is [called indirectly in `ReactReconciler`](https://github.com/facebook/react/blob/master/src/renderers/shared/stack/reconciler/ReactReconciler.js#L23-L25) in a helper function, `attachRefs`.

```js
function attachRefs() {
  ReactRef.attachRefs(this, this._currentElement);
}
```

That function [is enqueued in the `reactMountReady`](https://github.com/facebook/react/blob/master/src/renderers/shared/stack/reconciler/ReactReconciler.js#L66) callback, which is a pooled instance of `CallbackQueue`.

```js
transaction.getReactMountReady().enqueue(attachRefs, internalInstance);
```

The issue is that `CallbackQueue` only lets you provide the callback and the context for the callback; there's no way to pass any additional arguments to the enqueued callback. We *could* just tack on the `mockConfig` to `internalInstance` but then we'd need to add `_mockConfig` as a property on all instance types beforehand since at this point the instance disallows extensions, and I don't think `ReactReconciler` should be concerned with a transaction-specific config field.

Another option is to allow an argument (or multiple?) to be provided in `CallbackQueue`

```js
  enqueue: function(callback, context, arg) {
    this._callbacks = this._callbacks || [];
    this._contexts = this._contexts || [];
    this._args = this._args || [];
    this._callbacks.push(callback);
    this._contexts.push(context);
    this._args.push(arg);
  },
```
and then just pass in the argument when when we call `call` on the queued callback.

```js
      for (var i = 0; i < callbacks.length; i++) {
        callbacks[i].call(contexts[i], args[i]);
      }
```

cc @spicyj @gaearon what do you guys think?

Aweary commented:
@gaearon I've updated so it now uses the transaction to host the `mockConfig`. Instead of passing the entire Transaction to the `CallbackQueue` I just passed `mockConfig` directly and it seems to work well.

The only issue is legacy refs. They are currently attached to the instance directly, so if we're returning a mock instance the refs won't actually be accessible on the component via `this.refs` so it's a little tricky.

gaearon commented:
I don't quite understand, can you explain with code?

Aweary commented:
Writing it all out just made me realize that I was wrong 😄 

I was returning the mock instance for the component that the ref was being attached *to*, not the component/element that was being attached. It should be doable.

Aweary commented:
@gaearon updated with support for legacy refs as well

gaearon commented on src/renderers/shared/stack/reconciler/ReactCompositeComponent.js:
```diff
@@ -1100,10 +1100,17 @@ var ReactCompositeComponent = {
     invariant(inst != null, 'Stateless function components cannot have refs.');
-    var publicComponentInstance = component.getPublicInstance();
+    var publicComponentInstance;
+    if (mockConfig && mockConfig.getMockRef) {
```
`ReactCompositeComponent` should not be aware of test renderer-specific APIs. Why is defining `getPublicInstance()` inside test renderer not enough?

Aweary replied:
`getPublicInstance` is already defined inside the test renderer, but I didn't think to pass the `transaction` itself to `getPublicInstance`, which should let us clean up all this domain-specific code 👍

Aweary replied:
@gaearon one issue is that `component.getPublicInstance()` will be calling `ReactCompositeComponent.getPublicInstance` if the component is a composite. Even if I pass `transaction` all the way to `ReactCompositeComponent`, it would have to be aware of `transaction.mockConfig` if we wanted composite components to return mock instances.

gaearon replied:
>one issue is that component.getPublicInstance() will be calling ReactCompositeComponent.getPublicInstance if the component is a composite.

I think it’s reasonable to scope `getMockRef` effect to host instances only. You don’t want to mock composite refs, I think.

gaearon commented on src/renderers/testing/ReactTestRenderer.js:
```diff
@@ -36,13 +36,21 @@ function getRenderedHostOrTextFromComponent(component) {
   this._currentElement = element;
   this._renderedChildren = null;
   this._topLevelWrapper = null;
+  this._mockConfig = defaultMockConfig;
```
Can you please make it always read the mock config from transaction? By providing a default, I meant doing it at top level. So if you don’t pass `{ getMockRef }` to `ReactTestRenderer.create`, you get a default implementation *at that point*.

Aweary replied:
Sorry, I didn't remove this when I implemented the transaction code.

gaearon commented on src/renderers/testing/ReactTestMount.js:
```diff
@@ -135,24 +143,27 @@ ReactTestInstance.prototype.toJSON = function() {
+      mockConfig,
     );
-    return new ReactTestInstance(instance);
+    return new ReactTestInstance(instance, mockConfig);
```
I don't see this argument being used.

gaearon commented on src/renderers/shared/stack/reconciler/ReactRef.js:
```diff
@@ -16,15 +16,25 @@ var ReactOwner = require('ReactOwner');
 var ReactRef = {};
 
-function attachRef(ref, component, owner) {
+function attachRef(ref, component, owner, mockConfig) {
```
This is reconciler code, it also shouldn't have any knowledge about “mock config” (which is test renderer specific). We should *always* call `component.getPublicInstance()`. Passing an additional thing to it is fine, but that thing should also be renderer-agnostic (like `transaction`).

Instead of branching here, we should branch inside `getPublicInstance()` implementation in the test renderer. It also shouldn’t check if `mockConfig` exists because it should always exist—as we wanted to pass down a default value when user doesn’t specify any.

gaearon commented on src/renderers/shared/stack/reconciler/ReactReconciler.js:
```diff
@@ -20,8 +20,12 @@ var warning = require('warning');
  */
-function attachRefs() {
-  ReactRef.attachRefs(this, this._currentElement);
+function attachRefs(mockConfig) {
```
Same here—argument should be renderer-agnostic, `transaction` seems fine.

gaearon commented:
Left a few comments. General theme: reconciler should not have any code that is related only to test renderer. We should make keep any `getMockRef` calls and `mockConfig` access inside `getPublicInstance` of test renderer. Also let’s provide a default at the top level call level instead of initializing `_mockConfig` variable to a default in the constructor (I also couldn’t find where that is supposed to get overriden).

Aweary commented:
@gaearon thanks for the great feedback, passing the transaction to `getPublicInstance` was something I hadn't thought of that made the code a lot cleaner. I've updated the PR to address your comments 👍

gaearon commented on src/renderers/testing/ReactTestMount.js:
```diff
@@ -135,22 +150,25 @@ ReactTestInstance.prototype.toJSON = function() {
       batchedMountComponentIntoNode,
-      instance
+      instance,
+      mockConfig,
```
Can we pass default here instead, as this is the first place where it appears?

gaearon commented on src/renderers/testing/ReactTestReconcileTransaction.js:
```diff
@@ -57,9 +57,10 @@ var TRANSACTION_WRAPPERS = [ON_DOM_READY_QUEUEING];
  */
-function ReactTestReconcileTransaction() {
+function ReactTestReconcileTransaction(mockConfig) {
+  this._mockConfig = mockConfig;
```
Let’s keep it without underscore for consistency.
Also let’s move assignment just before `reactMountReady` assignment.

gaearon commented on src/renderers/testing/ReactTestReconcileTransaction.js:
```diff
@@ -57,9 +57,10 @@ var TRANSACTION_WRAPPERS = [ON_DOM_READY_QUEUEING];
+  this._mockConfig = mockConfig;
   this.reinitializeTransaction();
-  this.reactMountReady = CallbackQueue.getPooled(null);
+  this.reactMountReady = CallbackQueue.getPooled(this);
```
Since `this.reactMountReady` is accessed via `getReactMountReady()`, let’s add `getMockConfig()` for consistency and use that instead outside.

gaearon commented on src/renderers/testing/__tests__/ReactTestRenderer-test.js:
```diff
@@ -231,6 +232,86 @@ describe('ReactTestRenderer', () => {
+      </div>,
+      { getMockRef, foobar: true },
+    );
+
```
Let’s not leave more than one empty line.

gaearon commented on src/renderers/testing/__tests__/ReactTestRenderer-test.js:
```diff
@@ -231,6 +232,86 @@ describe('ReactTestRenderer', () => {
+
+    ReactTestRenderer.create(
+      <Foo />,
+      { getMockRef },
```
Style nit: `{getMockRef}` to be consistent with the codebase

gaearon commented on src/renderers/testing/__tests__/ReactTestRenderer-test.js:
```diff
@@ -231,6 +232,86 @@ describe('ReactTestRenderer', () => {
+    ]);
+  });
+
+
```
Same, extra newline

gaearon commented on src/renderers/testing/__tests__/ReactTestRenderer-test.js:
```diff
@@ -203,7 +204,7 @@ describe('ReactTestRenderer', () => {
     var log = [];
     ReactTestRenderer.create(<div ref={(r) => log.push(r)} />);
-    expect(log).toEqual([null]);
+    expect(log).toEqual([{}]);
```
Per @spicyj, let’s keep `null` the default one.

gaearon commented:
Looking really good, I hope you enjoyed the deep dive in our fettuccine.
A few more nits and I think this is good to go.

Aweary commented:
Thanks @gaearon, as always I appreciate all the guidance. I've addressed your comments.

>Looking really good, I hope you enjoyed the deep dive in our fettuccine.

It's been a lot of fun, I feel a lot more comfortable with the reconciliation code for sure 👍

gaearon commented on src/renderers/testing/ReactTestMount.js:
```diff
@@ -135,22 +148,25 @@ ReactTestInstance.prototype.toJSON = function() {
   render: function(
-    nextElement: ReactElement<any>
+    nextElement: ReactElement<any>,
+    mockConfig: TestRendererMockConfig,
```
Probably `?TestRendererMockConfig`, it’s optional.

gaearon commented on src/renderers/testing/ReactTestMount.js:
```diff
@@ -135,22 +148,25 @@ ReactTestInstance.prototype.toJSON = function() {
     var nextWrappedElement = React.createElement(
       TopLevelWrapper,
-      { child: nextElement }
+      { child: nextElement },
```
Nit: no need for extra spacing.

gaearon commented on src/renderers/testing/ReactTestMount.js:
```diff
@@ -135,22 +148,25 @@ ReactTestInstance.prototype.toJSON = function() {
     // The initial render is synchronous but any updates that happen during
     // rendering, in componentWillMount or componentDidMount, will be batched
     // according to the current batching strategy.
+    //
```
Nit: extra line

gaearon commented on src/renderers/testing/ReactTestRenderer.js:
```diff
@@ -60,12 +62,12 @@ ReactTestComponent.prototype.receiveComponent = function(
-  // I can't say this makes a ton of sense but it seems better than throwing.
-  // Maybe we'll revise later if someone has a good use case.
-  return null;
+ReactTestComponent.prototype.getPublicInstance = function(transaction) {
```
Let's add a newline before this line too for consistency now that we’ve put some of them between methods.

Aweary replied:
We're already calling `Object.assign` on the prototype, do you think it would be worth just assigning these there instead, like we do in a few other places? It would be a bit cleaner.

gaearon replied:
Let's keep it the way it is. If we want to make this cleaner we should just use ES6 classes instead but this should be done separately. We should not mix stylistic and meaningful changes in the same PR.

gaearon commented on src/renderers/testing/ReactTestRenderer.js:
```diff
@@ -60,12 +62,12 @@ ReactTestComponent.prototype.receiveComponent = function(
 };
+
 ReactTestComponent.prototype.unmountComponent = function() {};
 ReactTestComponent.prototype.toJSON = function() {
```
Same here

gaearon commented on src/renderers/testing/ReactTestRenderer.js:
```diff
@@ -134,9 +136,9 @@ ReactComponentEnvironment.injection.injectEnvironment({
   replaceNodeWithMarkup: function() {},
 });
 
+
```
Let’s remove this newline

gaearon commented on src/renderers/testing/ReactTestRenderer.js:
```diff
@@ -134,9 +136,9 @@ ReactComponentEnvironment.injection.injectEnvironment({
 var ReactTestRenderer = {
   create: ReactTestMount.render,
-
   /* eslint-disable camelcase */
```
And keep this one, so the diff is smaller. In general better not to formatting changes to code you didn’t directly touch unless it’s horribly formatted

gaearon commented on src/renderers/testing/__tests__/ReactTestRenderer-test.js:
```diff
@@ -1,3 +1,4 @@
+
```
Extra newline

gaearon commented on src/renderers/testing/__tests__/ReactTestRenderer-test.js:
```diff
@@ -231,6 +232,84 @@ describe('ReactTestRenderer', () => {
     );
   });
 
+  it('allows an optional getMockRef function', () => {
```
Let’s also test that if you pass `{}` as options, `getMockRef` will be returning `null` (as opposed to throwing on `options.getMockRef()`).

Aweary replied:
Currently if you pass in anything as options to `create` it does assume that you've provided a `getMockRef` function and it will throw. How do you want to handle that? Should `getMockRef` be required if you're passing in your own options? 

I'm assuming somewhere down the line we provide more user-configurable options to pass here, so I don't want to default to `defaultMockConfig` if `getMockRef` doesn't exist.

gaearon replied:
The way "options" are generally handled in JS libraries, user-passed options are shallowly merged with defaults, and the result is passed down. So passing `null`, `{}` or `{unrelatedOption:42}` gives you default `getMockRef` implementation.

Can we do the same here? `options = Object.assign({}, options, defaultOptions)`

Aweary commented:
@gaearon it's a little confusing that the ReactTestRenderer file defines three component types in one file (`ReactTestComponent`, `ReactTestTextComponent`, `ReactTestEmptyComponent`.

We have modules in other places that just export empty component types like `ReactDOMEmptyComponent`. Could we just do the same and move these into their own file to clean up the ReactTestRenderer file a bit?

gaearon commented:
We could do that but as a separate PR. Otherwise if other PRs change those files but we find a problem with this PR and need to revert it, we can't easily do that. Refactorings and features need to be submitted separately for this reason.

gaearon commented on src/renderers/testing/ReactTestMount.js:
```diff
@@ -149,24 +149,22 @@ var ReactTestMount = {
   render: function(
     nextElement: ReactElement<any>,
-    mockConfig: TestRendererMockConfig,
+    mockConfig?: TestRendererMockConfig,
```
I think we put `?` before the type elsewhere, like ?Stuff. Not sure if there's two ways of doing the same thing?

edvinerikson replied:
Adding `?` at the end of a object/argument key means that the field is optional while adding `?` in the beginning of the type means that it can be null or the type you specified however you always need to pass the argument/key.

```js
/* @flow */

function foo1(a?: { bar: number}) {
}

foo1(); // fine
foo1({}) // not fine

function foo2(a: { bar: ?number }) {
}

foo2({}); // invalid
foo2({ bar: 1}); // fine
foo2({ bar: null}); // fine
```

https://flowtype.org/try/#0PQKgBAAgZgNg9gdzCYAoVUCuA7AxgFwEs5swo44BGACgEMB+ALjAG8wAjWgJ2e0wFt2AUy4BfAJStUo9OSrVxAbjDBgZQtiEYKNFhJVrscfOs2ycBYqTkAmOszaceYen0Eiw+ltNkU7epQMwDQA3WhhCABNtOH8ObmZKCWVVUy1bakcEsD4YGGSgqA0hIA

gaearon commented on src/renderers/testing/ReactTestMount.js:
```diff
@@ -20,6 +20,16 @@ var getHostComponentFromComposite = require('getHostComponentFromComposite');
+  getMockRef: (element: ReactElement) => Object,
+};
+
+var defaultMockConfig = {
```
Let's also rename "mock config" to "test options" everywhere. It's longer but more explicitly and makes it clear we can add more stuff to it if we want to that's unrelated to mocking.

Inside test renderer we can just call it `options`. Outside, `testOptions` and `getTestOptions` seem fine.

gaearon commented on src/renderers/testing/ReactTestRenderer.js:
```diff
@@ -60,13 +62,15 @@ ReactTestComponent.prototype.receiveComponent = function(
-  return null;
+
+ReactTestComponent.prototype.getPublicInstance = function(transaction) {
+  return transaction.getMockConfig().getMockRef(this._currentElement);
```
Let's make this three lines. First get the element, then get the options, then use them to get and return the ref.

Aweary commented:
@gaearon I *think* I've addressed all your feedback so far, let me know if there's anything else you'd like to me work on/change.

gaearon commented:
My only concern is if `getMockRef` makes sense as a name. I don't have better suggestions though.

I'll mark this as accepted but I'd appreciate a comment from @spicyj before merge.

gaearon commented:
Maybe `getRefMock` is better? Or just `getMock`?

spicyj commented:
How about `createRefMock` or `createNodeMock`?

gaearon commented:
I didn’t want to say `Node` because we also use test renderer on native, and presumably at some point we’ll make native host instances have string element type? I don’t really know enough about RN to say.

`createRefMock()` sounds good to me. @Aweary let’s change to that and then get it in?

spicyj commented:
I was thinking of those as nodes too. :)

gaearon commented:
Hmm, then I like `createNodeMock` the most. Ref is more jargon-y and can be confused with composite refs.

Aweary commented:
@gaearon @spicyj updated it to `createNodeMock` 👍

gaearon commented:
Great job.
Leaving the honor of pressing "Squash and Merge" to you :smile:
